src/luarocks/fetch.lua

```diff
@@ -355,7 +356,29 @@ function fetch.get_sources(rockspec, extract, dest_dir)
       if not ok then return nil, err end
       fs.unpack_archive(rockspec.source.file)
       if not fs.exists(rockspec.source.dir) then
-         return nil, "Directory "..rockspec.source.dir.." not found inside archive "..rockspec.source.file, "source.dir", source_file, store_dir
+
+         -- If rockspec.source.dir can't be found, see if we only have one
+         -- directory in store_dir.  If that's the case, assume it's what
+         -- we're looking for.
+         -- We only do this if the rockspec source.dir was not set, and only
+         -- with rockspecs newer than 3.0.
+         local dir_count, found_dir = 0
+
+         if not rockspec.source.dir_set and deps.format_is_at_least(rockspec, "3.0") then
+            local files = fs.list_dir()
+            for _, f in ipairs(files) do
```

Comment:
Rethinking this, I believe it would be best to do something like

```lua
   if #files == 1 and fs.is_dir(files[1]) then
      found_dir = f
   end
```

In other words, assume the single directory is the `source.dir` only if it is the only file in the archive. If the archive has multiple top-level files, then `.` would probably be the best `source.dir` (archive does not contain a top-level directory and it's just a bunch of files in the root, typical .zip)